### PR TITLE
Timecode struct update for Milliseconds

### PR DESCRIPTION
**Pull request by Damien-Renault**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,33 +1,69 @@
...
+      self.hours, self.minutes, self.seconds, fraction
     )
   }
 }
 
+impl From<Duration> for Timecode {
+  fn from(duration: Duration) -> Self {
+    let hours = duration.as_secs() / 3600;
+    let minutes = duration.as_secs() / 60 - hours * 60;
+    let seconds = duration.as_secs() - hours * 3600 - minutes * 60;
```
Comment (VincentJousse) [inline src/lib.rs:51]:
```suggestion
    let remaining = duration.as_secs();
    let hours = remaining / 3600;
    let remaining %= 3600;
    let minutes = remaining / 60;
    let seconds = remaining % 60;
```

File: src/lib.rs [hunk 2]
```diff
@@ -1,33 +1,69 @@
...
+      hours: hours as u8,
+      minutes: minutes as u8,
+      seconds: seconds as u8,
+      fraction,
+    }
+  }
+}
+
 impl From<(u32, FrameRate)> for Timecode {
   fn from((frames, video_frame_rate): (u32, FrameRate)) -> Self {
```
Comment (VincentJousse) [inline src/lib.rs:68]: Rename video_frame_rate frame_rate

File: src/lib.rs [hunk 3]
```diff
@@ -110,6 +112,32 @@ impl Timecode {
...
+        frame_rate,
+      } => Frame {
+        frames,
+        drop_frame,
+        color_frame,
+        frame_rate,
+      },
+      MilliSeconds(ms) => MilliSeconds(ms),
+    }
+  }
```
Comment (VincentJousse) [inline src/lib.rs:140]: return a `&Fraction` and thus, just use &self.fraction

File: src/lib.rs [hunk 4]
```diff
@@ -1,16 +1,16 @@
-use crate::Fraction::{Frame, MilliSeconds};
 pub use frame_rate::FrameRate;
 use num_traits::cast::ToPrimitive;
 use serde::{Deserialize, Serialize};
-use std::string::ToString;
-use std::time::Duration;
+use std::{string::ToString, time::Duration};
+use Fraction::Frame;
+use Fraction::MilliSeconds;
```
Comment (VincentJousse) [inline src/lib.rs:6]: Remove these lines from imports

File: tests/parser.rs [hunk 5]
```diff
@@ -1,4 +1,7 @@
 extern crate timecode;
+
+use frame_rate::FrameRate::_25_00;
```
Comment (VincentJousse) [inline tests/parser.rs:3]: Import only frame_rate::FrameRate

**Review by VincentJousse — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 6]
```diff
@@ -1,60 +1,100 @@
...
 }
 
 impl ToString for Timecode {
   fn to_string(&self) -> String {
-    let separator = match self.drop_frame {
-      true => ';',
-      false => ':',
+    let fraction = match self.fraction {
+      Frame {
+        frames, drop_frame, ..
```
Comment (valnoel) [inline src/lib.rs:31]: Test

File: src/frame.rs [hunk 7]
```diff
@@ -0,0 +1,32 @@
+pub use frame_rate::FrameRate;
+use serde::{Deserialize, Serialize};
+#[derive(Debug, Deserialize, Clone, PartialEq, Serialize)]
```
Comment (VincentJousse) [inline src/frame.rs:3]: Jump a line after imports

File: src/frame.rs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+pub struct Frame {
```
Comment (VincentJousse) [inline src/frame.rs:4]: Rename struct Frame with TimecodeFrames
Rename file timecode_frames

File: src/frame.rs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  frames: u8,
```
Comment (VincentJousse) [inline src/frame.rs:5]: Rename number_of_frames

File: src/frame.rs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  }
+  pub fn color_frame(&self) -> bool {
+    self.color_frame
+  }
+  pub fn frame_rate(&self) -> FrameRate {
+    self.frame_rate
+  }
+  pub fn frames(&self) -> u8 {
+    self.frames
+  }
```
Comment (VincentJousse) [inline src/frame.rs:31]: Jump lines between functions

File: src/fraction.rs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  MilliSeconds(u16),
+}
+
+impl ToString for Fraction {
+  fn to_string(&self) -> String {
+    match self {
+      Fraction::Frame(frame) => {
+        let separator = if frame.drop_frame() { ';' } else { ':' };
+
+        format!("{}{:02}", separator, frame.frames())
```
Comment (VincentJousse) [inline src/fraction.rs:17]: Move this in frame.rs in :
```
impl ToString for Frame {
  fn to_string(&self) -> String {
  }
}
```

**Review by VincentJousse — CHANGES_REQUESTED** [review]

**Review by VincentJousse — CHANGES_REQUESTED** [review]

File: src/fraction.rs [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+#[derive(Debug, Deserialize, Clone, PartialEq, Serialize)]
+pub enum Fraction {
+  Frame(TimecodeFrames),
+  MilliSeconds(u16),
+}
+
+impl ToString for Fraction {
+  fn to_string(&self) -> String {
+    match self {
+      Fraction::Frame(..) => self.to_string(),
```
Comment (VincentJousse) [inline src/fraction.rs:14]:
```suggestion
      Fraction::Frames(timecode_frames) => timecode_frames.to_string(),
```

**Review by VincentJousse — CHANGES_REQUESTED** [review]

File: src/fraction.rs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+use super::TimecodeFrames;
+use serde::{Deserialize, Serialize};
+#[derive(Debug, Deserialize, Clone, PartialEq, Serialize)]
```
Comment (VincentJousse) [inline src/fraction.rs:3]: Jump a line

**Review by VincentJousse — CHANGES_REQUESTED** [review]

File: src/fraction.rs [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+pub enum Fraction {
+  Frames(TimecodeFrames),
+  MilliSeconds(u16),
+}
+
+impl ToString for Fraction {
+  fn to_string(&self) -> String {
+    match self {
+      Fraction::Frames(timecode_frame) => timecode_frame.to_string(),
```
Comment (VincentJousse) [inline src/fraction.rs:13]: Use Self::

File: src/fraction.rs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+      Fraction::MilliSeconds(milliseconds) => format!(".{:03}", milliseconds),
```
Comment (VincentJousse) [inline src/fraction.rs:14]: Use Self::

File: src/lib.rs [hunk 16]
```diff
@@ -1,75 +1,89 @@
...
+    let remaining = duration.as_secs();
+    let hours = remaining / 3600;
+    let remaining = remaining % 3600;
+    let minutes = remaining / 60;
+    let seconds = remaining % 60;
+    let ms = duration.as_millis() as u64
+      - hours * 36e5 as u64
+      - minutes * 6e4 as u64
+      - seconds * 1e3 as u64;
+    let fraction = Fraction::MilliSeconds(ms as u16);
```
Comment (VincentJousse) [inline src/lib.rs:42]:
```suggestion
    let fraction = Fraction::MilliSeconds(duration.subsec_millis() as u16);
```

File: src/lib.rs [hunk 17]
```diff
@@ -130,83 +144,166 @@ impl Timecode {
...
+    )
+  }
+}
+
+impl From<&Timecode> for f64 {
+  fn from(timecode: &Timecode) -> Self {
+    (timecode.hours as f64) * 3600.0 + (timecode.minutes as f64) * 60.0 + (timecode.seconds as f64)
+  }
+}
+
```
Comment (VincentJousse) [inline src/lib.rs:181]: Move unitary tests in a test module
(See https://doc.rust-lang.org/book/ch11-01-writing-tests.html)

File: src/lib.rs [hunk 18]
```diff
@@ -1,75 +1,89 @@
...
+}
 
-    format!(
-      "{:02}:{:02}:{:02}{}{:02}",
-      self.hours, self.minutes, self.seconds, separator, self.frame
-    )
+impl From<Duration> for Timecode {
+  fn from(duration: Duration) -> Self {
+    let remaining = duration.as_secs();
+    let hours = remaining / 3600;
```
Comment (VincentJousse) [inline src/lib.rs:34]: Make u8 conversion here

File: src/lib.rs [hunk 19]
```diff
@@ -1,75 +1,89 @@
...
+    let remaining = remaining % 3600;
+    let minutes = remaining / 60;
```
Comment (VincentJousse) [inline src/lib.rs:36]: Make u8 conversion here

File: src/lib.rs [hunk 20]
```diff
@@ -1,75 +1,89 @@
...
+    let seconds = remaining % 60;
```
Comment (VincentJousse) [inline src/lib.rs:37]: Make u8 conversion here

**Review by VincentJousse — CHANGES_REQUESTED** [review]

**Pull request merged**